### PR TITLE
Allow modification of existing functions

### PR DESCRIPTION
**Pull request by jimhester**
This simply adds a `...` argument to `argufy`, which allows you to modify existing functions by supplying the new argufy arguments.

e.g.

``` r
tolower <- argufy(base::tolower, x = ~ is.character)
tolower(1)
# Error: is.character(x) is not TRUE
```

This is a good idea! https://github.com/smbache/ensurer has a similar idea, but I like this API and implementation better.

It might be interesting to have this generate https://github.com/hadley/assertthat calls rather than `stopifnot()`, perhaps as an option?

### CODE REVIEW

**Comment by gaborcsardi** [comment]
Thanks! This looks great! 

A was thinking about how to best include hadley/assertthat, and indeed `stopifnot()` is just a placeholder for now. Btw. you can already use `assertthat` checks, you import them. If they don't start with `is.` or `is_`, then you need to write the full expression.

Anyway, I'll look this over tonight, need to run now. Btw. could you generate the Rd docs with https://github.com/gaborcsardi/maxygen? Thanks.

**Comment by jimhester** [comment]
Ah I see you thought the same about `assertthat`, did not see #1!

I regenerated docs with maxygen (another good idea! I have long wished for markdown in roxygen).  Let me know if there is anything else I need to change.

File: NAMESPACE [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Generated by roxygen2 (4.1.1.9000): do not edit by hand
+# Generated by roxygen2 (4.1.1): do not edit by hand
```
Comment (gaborcsardi) [inline NAMESPACE:1]: Can you revert this please?

File: R/package.R [hunk 2]
```diff
@@ -37,15 +39,27 @@
 #' })
 #'
 #' prefix
+#'
+#' # modify tolower to fail if given a non-character
+#' tolower <- argufy(base::tolower, x = ~ is.character)
```
Comment (gaborcsardi) [inline R/package.R:44]: Maybe if you do `as.character` instead, then the example runs. Not a big deal, though.

File: man/as_enum.Rd [hunk 3]
```diff
@@ -1,4 +1,4 @@
-% Generated by roxygen2 (4.1.1.9000): do not edit by hand
+% Generated by roxygen2 (4.1.1): do not edit by hand
```
Comment (gaborcsardi) [inline man/as_enum.Rd:1]: Can you please revert this, too? Thanks.

**Comment by gaborcsardi** [comment]
Commented on some minor things. It looks pretty great! 
Can you please squash? Then I'll merge it.

Also, please let me know what you think about the assertthat "merge"? I'll add some ideas there.

File: R/package.R (hunk 2 above)
Comment (jimhester) [inline R/package.R:44]: True, I guess was trying to show it Error to piece the modification
worked.  Could have surrounded with testthat::expect_error() I suppose...

In R/package.R
https://github.com/gaborcsardi/argufy/pull/4#discussion_r36694138:

> @@ -37,15 +39,27 @@
>  #' })
>  #'
>  #' prefix
> +#'
> +#' # modify tolower to fail if given a non-character
> +#' tolower <- argufy(base::tolower, x = ~ is.character)

Maybe if you do as.character instead, then the example runs. Not a big
deal, though.

—
Reply to this email directly or view it on GitHub
https://github.com/gaborcsardi/argufy/pull/4/files#r36694138.

Comment (gaborcsardi) [inline R/package.R:44]: Yeah, but that loads testthat. I think the `\dontrun` is actually fine. We should really have a way to have errors in examples. Examples are not test cases..... anyway, just leave it as it is.

**Comment by gaborcsardi** [comment]
Actually, it is fine as it is. :) Thanks again.

**Comment by gaborcsardi** [comment]
Btw. I knew about ensurer, but didn't know about https://github.com/smbache/ensurer#types-and-type-safe-functions-new-experimental-subject-to-change Very similar! It is also amazing that we both chose the tilde. I guess there aren't too many choices, I was thinking about `?` first....

**Pull request merged**